File: week 3/change/change.cpp
#include <iostream>
#include <cassert>

int get_change(int m) {
  int denom[] = {1, 5, 10};
  int index = (sizeof(denom) / sizeof(int)) - 1;
  int total = 0;

  while (m > 0) {
    if (m >= denom[index]) {
      total += m / denom[index];
      m %= denom[index];
    } else {
      index -= 1;
    }
  }

  return total;
}

void test_solution() {
  assert(get_change(0) == 0);
  assert(get_change(10) == 1);
  assert(get_change(9) == 5);
  assert(get_change(2) == 2);
  assert(get_change(28) == 6);
}

int main() {
  test_solution();

  int m;
  std::cin >> m;
  std::cout << get_change(m) << '\n';

  return 0;
}

File: week 3/change/change.py
# Uses python3
import sys
import math

def get_change(m):
    denom = [1, 5, 10]
    index = len(denom) - 1
    total = 0

    while m > 0:
        if m >= denom[index]:
            total += math.floor(m / denom[index])
            m %= denom[index]
        else:
            index -= 1

    return total

def test_solution():
    assert get_change(0) == 0
    assert get_change(10) == 1
    assert get_change(9) == 5
    assert get_change(2) == 2
    assert get_change(28) == 6

if __name__ == '__main__':
    test_solution()

    while True:
        print get_change(int(sys.stdin.readline()))
